### ISSUE TITLE
`PrivateKey.derive()` and/or `PrivateKey.legacyDerive()` not translating well from v1 SDK

### ISSUE THREAD

**Issue by robkainz** [hashgraph/hedera-sdk-js#417]
The error "(BUG) keyToProtobuf: unsupported key type: Ed25519PublicKey" is thrown while attempting to execute a TokenCreateTransaction in a React Native app.

Additionally, in order to get the app to start cleanly, I had to explicitly add the following dependencies to my project due errors similar to this one:

> error: Error: Unable to resolve module `@stablelib/utf8` from `node_modules/@hashgraph/sdk/lib/crypto/Keystore.js`: @stablelib/utf8 could not be found within the project.

Additional dependencies:

```
"@improbable-eng/grpc-web": "^0.14.0",
"@stablelib/base64": "^1.0.1",
"@stablelib/hex": "^1.0.1",
"@stablelib/hmac": "^1.0.1",
"@stablelib/pbkdf2": "^1.0.1",
"@stablelib/sha256": "^1.0.1",
"@stablelib/sha384": "^1.0.1",
"@stablelib/utf8": "^1.0.1",
"google-protobuf": "^3.17.3",

```
Here's the code used to create the token transaction:

```
const privateKeyHex = jsonConfig.privateKey;
const privateKey:Ed25519PrivateKey = Ed25519PrivateKey.fromString(privateKeyHex);
const operatorPublicKey:PublicKey = privateKey.publicKey;

try {
    transactionResponse = await new TokenCreateTransaction()
        .setTokenName(name)
        .setTokenSymbol(symbol)
        .setMaxTransactionFee(maxTransactionFeeHbar)
        .setDecimals(decimalsBigNumber.toNumber()) 
        .setInitialSupply(initialSupplyBigNumber.toNumber()) 
        .setTreasuryAccountId(jsonConfig.accountId)
        .setAdminKey(operatorPublicKey)
        .setFreezeKey(operatorPublicKey)
        .setWipeKey(operatorPublicKey)
        .setKycKey(operatorPublicKey)
        .setSupplyKey(operatorPublicKey)
        .setFreezeDefault(false)
        .execute(client);
}
catch(err) {
    throw err;
}
```

**Comment by danielakhterov** [comment]
I'm a bit confused, are you using the v1 SDK somehow? v2 SDK does not have `Ed25519PrivateKey` we've replaced that with `Privatekey`

**Comment by robkainz** [comment]
Hey Daniel,

I could not figure out how to get the expected keypair using PrivateKey, and fortunately for me the Ed25519PrivateKey still worked with the right import statement. I can confirm I am using only the following with a clean install:

> "@hashgraph/sdk": "^2.0.22", 

The version using Ed25519PrivateKey will give me a public key (using derive2(0)) that matches the one the Ledger Nano provides. The new API does not as far as I have been able to tell. Please tell me if I am missing something or doing something incorrectly. I am not sure yet whether it's a mistake on my part or a bug in the SDK. 

Here's an example:

> mnemonic: tray trust alert balance rubber seat own use critic protect crucial license

I used this import:

`import` {Ed25519PrivateKey} from "@hashgraph/sdk/lib/crypto/Ed25519PrivateKey";

This method gives me the public key f3aec5b29a666a80c42652bc8528e2e831ba54fa0053aa39e4d9bd242ef281fc (which is what I expect and matches my previous golang implementation and will match what the Nano gives:

```
export async function mnemonicToPublicKey(mnemonic:string, passphrase = "") {
    const m:any = await Mnemonic.fromString(mnemonic);
    let pk:Ed25519PrivateKey = await Ed25519PrivateKey.fromMnemonic(m, passphrase);
    pk = await pk.derive2(0);
    const rawPublicKey:string = Buffer.from(pk.publicKey.toBytes()).toString('hex');
    console.log('rawPublicKey: ', rawPublicKey) //prints f3aec5b29a666a80c42652bc8528e2e831ba54fa0053aa39e4d9bd242ef281fc
    return rawPublicKey;
}
```

This version gives me an entire different public key:

```
export async function mnemonicToPublicKeyINCORRECT(mnemonic, passphrase = "") {
    const m:Mnemonic = await Mnemonic.fromString(mnemonic);
    const pk:PrivateKey = await m.toPrivateKey(passphrase);
    const schemePk:PrivateKey = await pk.derive(0);
    const rawPublicKey:string = Buffer.from(schemePk.publicKey.toBytes()).toString('hex');
    console.log('rawPublicKey: ', rawPublicKey) //prints: 63f8ca3239d20b93e2c3ad212e966736636654d63ce136a4219f2a7016008104
    return rawPublicKey;
}
```

I would love to know what I am doing wrong with the latter version.

**Comment by danielakhterov** [comment]
Have you tried `derive(0xFF)`? I'll add a test for this, but this might be a place where a certain `index` needs to be provided for `derive()` to produce the same key.
Going to hijack this issue to test this.

**Comment by danielakhterov** [comment]
@robkainz Did some quick testing, it's not obvious and we should really clear up the documentation, but to produce the same key in v2 you need to use `PrivateKey.derive(-1)`. Unfortunately, I do not remember exactly why this is the case.

**Comment by robkainz** [comment]
@danielakhterov Thank you for the quick response and suggested solution. I will try this and hopefully it will work, but just for the record, index 0 worked as expected for v2.1.3 of the API for the golang SDK. Ideally, IMHO, "derive(0)" would behave the same across all Hedera SDKs and because it doesn't, it means I now need to provide more extensive unit tests for this (and possibly other Hedera SDK keypair-related functionality?) that honestly should be outside the scope of my application development and should be handled by the SDK unit test suite instead. I personally think better care should be taken such that the interfaces to critical, keypair-related math functions are implemented uniformly across all Hedera SDKs to help inspire confidence to those developers, like myself, who have now used 3 or more Hedera SDKs.

```
 func MnemonicToPublicKey(mnemonic string, passphrase string) (string, error) {
 	m, err := hedera.MnemonicFromString(mnemonic)
 	if err!=nil {
 		return "", err
 	}
 	pk, err := hedera.PrivateKeyFromMnemonic(m, passphrase)
 	if err!=nil {
 		return "", err
 	}

	pk, err = pk.Derive(0)
	if err!=nil {
		return "", err
	}

 	raw := pk.PublicKey().Bytes()
 	rawEncoded := hex.EncodeToString(raw)
 	return rawEncoded, nil
 }
```

**Comment by danielakhterov** [comment]
You're absolutely right. This is an area we could and should improve on. I completely agree that an index of `0` working on one SDK, but requires index of `-1` in another is sub-optimal to say the least.

**Comment by robkainz** [comment]
Hey @danielakhterov ,

I'm afraid your fix did not work. I modified my function to use the index -1 as you suggested, but I am still not getting the correct public key for this mnemonic phrase. Please re-open this defect.

```
export async function mnemonicToPublicKey(mnemonic, passphrase = "") {
    const m:Mnemonic = await Mnemonic.fromString(mnemonic);
    const pk:PrivateKey = await m.toPrivateKey(passphrase);
    const schemePk:PrivateKey = await pk.derive(-1);
    const rawPublicKey:string = Buffer.from(schemePk.publicKey.toBytes()).toString('hex');
    console.log('rawPublicKey: ', rawPublicKey)
    return rawPublicKey;
}
```

Here is the output I got:

> [Mon Jun 28 2021 07:55:06.311]  LOG      rawPublicKey:  6ab342c5ef89b275477d99c56554076d1eb293605fb65b2b3d11fed02d48f047

The expected value for derive(0) is: 

> f3aec5b29a666a80c42652bc8528e2e831ba54fa0053aa39e4d9bd242ef281fc

for the mnemonic:

> tray trust alert balance rubber seat own use critic protect crucial license

**Comment by mehcode** [comment]
> I also just noticed with the latest version of the SDK (2.0.23), access to the Ed25519PrivateKey is gone. This means until this is fixed, my only option is to run with the 2.0.22 SDK using that Ed25519PrivateKey hack (short of going back to golang, causing me a lot of additional work.)

I'm going to jump in and see what I can figure out but I don't understand this comment. `Ed25519PrivateKey` is a v1 class and has never existed in v2. What are we talking about here?
